### PR TITLE
fix(sessds): Fix crash resulting from subscription state update

### PR DESCRIPTION
**Pull request by ieQu1**
Fixes: #14039

Release version: v/e5.8.2

## Summary
Properly handle change of the subscription parameters in the durable sessions.

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [ ] Added property-based tests for code which performs user input validation
- [ ] Changed lines covered in coverage report
- [x] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [x] For internal contributor: there is a jira ticket to track this change
- [ ] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket
- [ ] Schema changes are backward compatible

## Checklist for CI (.github/workflows) changes

- [ ] If changed package build workflow, pass [this action](https://github.com/emqx/emqx/actions/workflows/build_packages.yaml) (manual trigger)
- [ ] Change log has been added to `changes/` dir for user-facing artifacts update

### CODE REVIEW

File: apps/emqx/test/emqx_persistent_session_ds_SUITE.erl [hunk 1]
```diff
@@ -413,6 +413,76 @@ t_session_unsubscription_idempotency(Config) ->
...
+            ?assertMatch(
+                #{
+                    SSID1 := #{subopts := #{qos := ?QOS_1}, superseded_by := SSID2},
+                    SSID2 := #{subopts := #{qos := ?QOS_2}}
+                },
+                SStates2
+            ),
+            %% Now ack the packet and publish some more to trigger SRS
+            %% update. This should, in effect, release the reference
+            %% to the old subscription state, and let GC will delete
```
Comment (ieQu1) [inline apps/emqx/test/emqx_persistent_session_ds_SUITE.erl:469]:
```suggestion
            %% to the old subscription state, and let GC delete
```

File: apps/emqx/test/emqx_persistent_session_ds_SUITE.erl [hunk 2]
```diff
@@ -413,6 +413,76 @@ t_session_unsubscription_idempotency(Config) ->
...
+            %% Subscribe to the topic using QoS1 initially:
+            {ok, _, _} = emqtt:subscribe(Sub, TopicFilter, ?QOS_1),
+            #{subscriptions := Subs1} = GetS(),
+            #{TopicFilter := #{current_state := SSID1}} = Subs1,
+            %% Fill the storage with some data to create the streams:
+            {ok, _} = emqtt:publish(Pub, <<"t/1">>, <<"1">>, ?QOS_2),
+            [#{packet_id := PI1, qos := ?QOS_1}] = emqx_common_test_helpers:wait_publishes(
+                1, 5_000
+            ),
+            %% Upgrade subscription to QoS2 and wait for the session GC
```
Comment (ieQu1) [inline apps/emqx/test/emqx_persistent_session_ds_SUITE.erl:452]:
```suggestion
            %% Upgrade subscription to QoS2 and wait for the session-internal GC
```

File: apps/emqx/test/emqx_persistent_session_ds_SUITE.erl [hunk 3]
```diff
@@ -413,6 +413,76 @@ t_session_unsubscription_idempotency(Config) ->
...
 
+%% This testcase verifies that the session handles update of the
+%% subscription settings by the client correctly.
+t_subscription_state_change(Config) ->
+    [Node1Spec | _] = ?config(node_specs, Config),
+    [Node1] = ?config(nodes, Config),
+    Port = get_mqtt_port(Node1, tcp),
+    TopicFilter = <<"t/+">>,
+    ClientId = mk_clientid(?FUNCTION_NAME, sub),
+    %% Helper function that waits for the session GC:
```
Comment (ieQu1) [inline apps/emqx/test/emqx_persistent_session_ds_SUITE.erl:424]:
```suggestion
    %% Helper function that waits for the internal session GC:
```

**Review by keynslug — APPROVED** [review]

**Review by thalesmg — APPROVED** [review]

**Comment by emqxqa** [comment]
[TestExecution](https://emqx.atlassian.net/browse/QA-4796)

**Pull request merged**